File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution{
    public int[] nodesBetweenCriticalPoints(ListNode head){
        
        List<Integer> temp=new ArrayList<>();
        ListNode prev=head;
        ListNode curr=head.next;
        int c=1;
        while(curr != null && curr.next != null){
            if(curr.val > prev.val && curr.val > curr.next.val){
                temp.add(c);
            }
            if(curr.val < prev.val && curr.val < curr.next.val){
                temp.add(c);
            }

            c++;
            prev=prev.next;
            curr=curr.next;
        }

        if(temp.size()<2)
            return new int[]{-1, -1};

        int mn=Integer.MAX_VALUE;
        int mx=temp.get(temp.size()-1)-temp.get(0);
        for(int i=0; i<temp.size()-1; i++){
            mn=Math.min(mn, temp.get(i+1)-temp.get(i));
        }

        return new int[]{mn, mx};
    }
}

File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        temp=[]
        prev=head
        curr=head.next
        c=1

        while curr is not None and curr.next is not None:
            if curr.val > prev.val and curr.val > curr.next.val:
                temp.append(c)
            if curr.val < prev.val and curr.val < curr.next.val:
                temp.append(c)

            c+=1
            prev=prev.next
            curr=curr.next

        if len(temp)<2:
            return [-1, -1]

        mn=float('inf')
        mx=temp[-1]-temp[0]
        for i in range(len(temp)-1):
            mn=min(mn, temp[i+1]-temp[i])

        return [mn, mx]